File: soln-tree-problems/least-common-ancestor-binary-search-tree.java
/**
 @author Farheen Bano
  
 References-
 https://www.geeksforgeeks.org/lowest-common-ancestor-binary-tree-set-1/
 https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/
*/

import java.util.*;
import java.lang.*;
import java.io.*;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        int parentVal=root.val;
        int pVal=p.val;
        int qVal=q.val;
        
        if(pVal<parentVal && qVal<parentVal){
            return lowestCommonAncestor(root.left,p,q);
        }
        else if(pVal>parentVal && qVal>parentVal){
            return lowestCommonAncestor(root.right,p,q);
        }
        else{
            return root;
        }
        
    }
}

File: C++/soln-tree-binary-problems/least-common-ancestor-binary-search-tree.cpp
/**
 @author Farheen Bano
  
 References-
 https://www.geeksforgeeks.org/lowest-common-ancestor-binary-tree-set-1/
 https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(p->val<root->val && q->val<root->val){
            return lowestCommonAncestor(root->left,p,q); 
        }
        else if(p->val>root->val && q->val>root->val){
            return lowestCommonAncestor(root->right,p,q); 
        }
        else
            return root;
    }
};